File: python/dfs/101_symmetric_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.check(root.left, root.right)

    def check(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p is None and q is None:
            return True
        if p and q and p.val == q.val:
            return self.check(p.left, q.right) and self.check(p.right, q.left)

        return False

File: go/dfs/101_symmetric_tree.go
package dfs

// Definition for a binary tree node.

func IsSymmetric(root *TreeNode) bool {
	return check(root.Left, root.Right)
}

func check(p *TreeNode, q *TreeNode) bool {
	if p == nil || q == nil {
		return p == q
	}
	if p.Val != q.Val {
		return false
	}
	return check(p.Left, q.Right) && check(p.Right, q.Left)
}
